### PR TITLE
Report HMR latency as trace spans for Turbopack

### PR DESCRIPTION
**Pull request by wbinnssmith**
This:
- Makes makes reporting hmr changes more accurate by emitting an explicit start event and lowering the aggregation period for reporting completed turbo tasks
- Parameterizes the aggregation period, allowing JS subscribers  to request different intervals, such as for the next-rs-api test
- Aligns path pattern with webpack by normalizing its layer names
- Fixes a bug in `drainAndGetNext` where the stream never could drain past the first item
- Sends `client-hmr-latency` spans from Turbopack in both page and app router pages

Closes PACK-2252

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7b526fb2a0d68e9526b5c7e0681a7267f0ca5976

`pnpm test test/integration/auto-export-query-error/test/index.test.js`
- Auto Export > production mode > should show warning for query provided for auto exported page correctly

<details>
<summary>Expand output</summary>

  ● Auto Export › production mode › should show warning for query provided for auto exported page correctly

    thrown: "Exceeded timeout of 60000 ms for a hook.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      23 | describe('Auto Export', () => {
      24 |   ;(process.env.TURBOPACK ? describe.skip : describe)('production mode', () => {
    > 25 |     beforeAll(async () => {
         |     ^
      26 |       const { stderr: curStderr, code: curCode } = await nextBuild(appDir, [], {
      27 |         stderr: true,
      28 |       })

      at beforeAll (integration/auto-export-query-error/test/index.test.js:25:5)
      at integration/auto-export-query-error/test/index.test.js:24:52
      at Object.describe (integration/auto-export-query-error/test/index.test.js:23:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7b526fb2a0d68e9526b5c7e0681a7267f0ca5976

`pnpm test test/integration/app-dir-export/test/dynamicapiroute-prod.test.ts`
- app dir - with output export - dynamic api route prod > production mode > should work in prod with dynamicApiRoute undefined
- app dir - with output export - dynamic api route prod > production mode > should work in prod with dynamicApiRoute 'error'
- app dir - with output export - dynamic api route prod > production mode > should work in prod with dynamicApiRoute 'force-static'
- app dir - with output export - dynamic api route prod > production mode > should work in prod with dynamicApiRoute 'force-dynamic'

<details>
<summary>Expand output</summary>

  ● app dir - with output export - dynamic api route prod › production mode › should work in prod with dynamicApiRoute undefined

    thrown: "Exceeded timeout of 60000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      12 |           'export const dynamic = "force-dynamic" on page "/api/json" cannot be used with "output: export".',
      13 |       },
    > 14 |     ])(
         |      ^
      15 |       'should work in prod with dynamicApiRoute $dynamicApiRoute',
      16 |       async ({ dynamicApiRoute, expectedErrMsg }) => {
      17 |         await runTests({ isDev: false, dynamicApiRoute, expectedErrMsg })

      at ../node_modules/.pnpm/jest-each@29.7.0/node_modules/jest-each/build/bind.js:47:15
          at Array.forEach (<anonymous>)
      at integration/app-dir-export/test/dynamicapiroute-prod.test.ts:14:6
      at integration/app-dir-export/test/dynamicapiroute-prod.test.ts:4:52
      at Object.describe (integration/app-dir-export/test/dynamicapiroute-prod.test.ts:3:1)

  ● app dir - with output export - dynamic api route prod › production mode › should work in prod with dynamicApiRoute 'error'

    Failed to replace content.

    Pattern: const dynamic = 'force-static'

    Content: export const dynamic = undefined

    export async function GET() {
      return Response.json({ answer: 42 })
    }

      669 |     } else if (typeof pattern === 'string') {
      670 |       if (!currentContent.includes(pattern)) {
    > 671 |         throw new Error(
          |               ^
      672 |           `Failed to replace content.\n\nPattern: ${pattern}\n\nContent: ${currentContent}`
      673 |         )
      674 |       }

      at File.replace (lib/next-test-utils.ts:671:15)
      at replace (integration/app-dir-export/test/utils.ts:118:13)
      at integration/app-dir-export/test/dynamicapiroute-prod.test.ts:17:23

  ● app dir - with output export - dynamic api route prod › production mode › should work in prod with dynamicApiRoute 'force-static'

    Failed to replace content.

    Pattern: const dynamic = 'force-static'

    Content: export const dynamic = undefined

    export async function GET() {
      return Response.json({ answer: 42 })
    }

      669 |     } else if (typeof pattern === 'string') {
      670 |       if (!currentContent.includes(pattern)) {
    > 671 |         throw new Error(
          |               ^
      672 |           `Failed to replace content.\n\nPattern: ${pattern}\n\nContent: ${currentContent}`
      673 |         )
      674 |       }

      at File.replace (lib/next-test-utils.ts:671:15)
      at replace (integration/app-dir-export/test/utils.ts:118:13)
      at integration/app-dir-export/test/dynamicapiroute-prod.test.ts:17:23

  ● app dir - with output export - dynamic api route prod › production mode › should work in prod with dynamicApiRoute 'force-dynamic'

    Failed to replace content.

    Pattern: const dynamic = 'force-static'

    Content: export const dynamic = undefined

    export async function GET() {
      return Response.json({ answer: 42 })
    }

      669 |     } else if (typeof pattern === 'string') {
      670 |       if (!currentContent.includes(pattern)) {
    > 671 |         throw new Error(
          |               ^
      672 |           `Failed to replace content.\n\nPattern: ${pattern}\n\nContent: ${currentContent}`
      673 |         )
      674 |       }

      at File.replace (lib/next-test-utils.ts:671:15)
      at replace (integration/app-dir-export/test/utils.ts:118:13)
      at integration/app-dir-export/test/dynamicapiroute-prod.test.ts:17:23

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 1]
```diff
@@ -549,30 +556,12 @@ async function startWatcher(opts: SetupOpts) {
...
-      // We've detected a change in some part of the graph. If nothing has
-      // been inserted into building yet, then this is the first change
-      // emitted, but their may be many more coming.
-      if (!hmrBuilding) {
-        hotReloader.send({ action: HMR_ACTIONS_SENT_TO_BROWSER.BUILDING })
-        hmrBuilding = true
-      }
+    function enqueueHmr(key: string, id: string, payload: HMR_ACTION_TYPES) {
       hmrPayloads.set(`${key}:${id}`, payload)
-      hmrEventHappened = true
```
Comment (sokra) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:561]: Shouldn't it keep that for the reporting to be printed?

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 2]
```diff
@@ -549,30 +556,12 @@ async function startWatcher(opts: SetupOpts) {
...
-      // been inserted into building yet, then this is the first change
-      // emitted, but their may be many more coming.
-      if (!hmrBuilding) {
-        hotReloader.send({ action: HMR_ACTIONS_SENT_TO_BROWSER.BUILDING })
-        hmrBuilding = true
-      }
+    function enqueueTurbopackMessage(payload: TurbopackUpdate) {
       turbopackUpdates.push(payload)
-      hmrEventHappened = true
-      sendHmrDebounce()
```
Comment (sokra) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:575]: Why no longer calling this? Shouldn't it send out the events immediately and don't wait for the end event?

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| buildDuration | 12.8s | 12.8s | N/A |
| buildDurationCached | 7.2s | 6.1s | N/A |
| nodeModulesSize | 200 MB | 200 MB | N/A |
| nextStartRea..uration (ms) | 421ms | 428ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.4 kB | 53.4 kB | ✓ |
| 433-HASH.js gzip | 29 kB | 29 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 240 B | 242 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.6 kB | 98.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.18 kB | 4.18 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| index.html gzip | 527 B | 528 B | N/A |
| link.html gzip | 539 B | 541 B | N/A |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 523 B | 523 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94 kB | 93.9 kB | N/A |
| page.js gzip | 148 kB | 148 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 627 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.5 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js wbinnssmith/timing | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 169 kB | 169 kB | ✓ |
| app-page-exp..prod.js gzip | 95.6 kB | 95.6 kB | ✓ |
| app-page-tur..prod.js gzip | 96.3 kB | 96.3 kB | ✓ |
| app-page-tur..prod.js gzip | 90.8 kB | 90.8 kB | ✓ |
| app-page.run...dev.js gzip | 142 kB | 142 kB | ✓ |
| app-page.run..prod.js gzip | 90.2 kB | 90.2 kB | ✓ |
| app-route-ex...dev.js gzip | 24.2 kB | 24.2 kB | N/A |
| app-route-ex..prod.js gzip | 16.8 kB | 16.8 kB | N/A |
| app-route-tu..prod.js gzip | 16.8 kB | 16.8 kB | N/A |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | N/A |
| app-route.ru...dev.js gzip | 23.6 kB | 23.6 kB | N/A |
| app-route.ru..prod.js gzip | 16.4 kB | 16.4 kB | N/A |
| pages-api-tu..prod.js gzip | 9.39 kB | 9.39 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.67 kB | 9.67 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.39 kB | 9.39 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 49.7 kB | 49.7 kB | N/A |
| Overall change | 779 kB | 779 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 8b358447ea07199e9804dce241b7bfa77e3dd819

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts (hunk 2 above)
Comment (sokra) [reply packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:575]: What's the answer to this?

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 3]
```diff
@@ -1748,6 +1731,33 @@ async function startWatcher(opts: SetupOpts) {
...
+        switch (updateMessage.updateType) {
+          case 'start': {
+            if (!hmrBuilding) {
+              hotReloader.send({ action: HMR_ACTIONS_SENT_TO_BROWSER.BUILDING })
+              hmrBuilding = true
+            }
+            break
+          }
+          case 'end': {
+            sendHmrDebounce()
```
Comment (sokra) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:1748]: I don't want to wait for `end` to receive hmr update.

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts (hunk 2 above)
Comment (wbinnssmith) [reply packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:575]: I reintroduced this call, so it's updating immediately again.

**Review by sokra — APPROVED** [review]

**Pull request merged**